### PR TITLE
textparse: Refactored benchmark

### PR DESCRIPTION
**Pull request by bwplotka**
For more realistic benchmarks to iterate over proto improvements we planned, https://github.com/prometheus/prometheus/pull/14965 and perhaps experimenting with CI use with https://github.com/benchmark-action/github-action-benchmark or so, I propose a refactor of the benchmark we want to persist in our `textparse` package.

To me we want to persist useful stuff to iterate on, but more importantly, something anyone should run before commiting new changes. For this reasons I killed those isolated cases for series, CT, series+metrics, series+metrics reuse. Instead we only have benchmark for realistic use of parser (as we use in `scrape.append`) AND opt-in (skipped by default) expfmt comparison. Anyone can comment out certain methods for local testing and comparisons.

Additionally:
* I provided example CLI command
* Fixed cases formatting to match official syntax (you can learn more in https://www.bwplotka.dev/2024/go-microbenchmarks-benchstat/).
* Killed unnecessary sinks. With interfaces and internal state, I don't see any difference if I have sinks or not (that total var).
* I changed operation to be = full scrape, not a single sample. I might miss why we did this per sample, but it feels unnecessary and complex. Each sample is different so perhaps it's simpler to think about ops = scrape (which we care about). 
* Removed gzip benchmark, I guess it was for some experiment, but not sure if worth to benchmark always not ours gzip implementation we don't implement.

Would love to know your opinion @beorn7 and @bboreham on this 🤗 

Here is an example of what this benchmark enables us to do https://github.com/prometheus/prometheus/pull/14965#issuecomment-2391021074

### CODE REVIEW

**Comment by bwplotka** [comment]
cc @krajorama as it will be useful for nhcb work too 🤗

File: model/textparse/benchmark_test.go [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+// Series, Series+Metrics with and without reuse, Series+CT. Those cases are sometimes
+// good to know if you are working on a certain optimization, but it does not
+// make sense to persist such cases for everybody (e.g. for CI one day).
+// For local iteration, feel free to adjust cases/comment out code etc.
+//
+// NOTE(bwplotka): Do not try to conclude "what parser (OM, proto, prom) is the fastest"
+// as the testdata has different amount and type of metrics and features (e.g. exemplars).
+func BenchmarkParse(b *testing.B) {
+	for _, bcase := range []struct {
+		dataFile  string // localized to ./testdata
```
Comment (krajorama) [inline model/textparse/benchmark_test.go:66]:
```suggestion
		dataFile  string // Localized to "./testdata".
```

File: model/textparse/benchmark_test.go [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+					_ = p.Metric(&res)
+					_ = p.CreatedTimestamp()
+					for hasExemplar := p.Exemplar(&e); hasExemplar; hasExemplar = p.Exemplar(&e) {
+					}
+				}
+			}
+		})
+
+		// Compare with expfmt, opt-in, no need to benchmark external code.
```
Comment (krajorama) [inline model/textparse/benchmark_test.go:151]: This comment is pretty confusing as by default we run this check for the two text parsers, so it's not opt-in.

File: model/textparse/promparse_test.go [hunk 3]
```diff
@@ -487,190 +483,3 @@ func TestPromNullByteHandling(t *testing.T) {
...
-						decSamples = decSamples[:0]
-					}
-				}
-				_ = total
-			})
-		}
-	}
-}
-
-func BenchmarkGzip(b *testing.B) {
```
Comment (krajorama) [inline model/textparse/promparse_test.go:638]: nit: worth a mention in the PR description as we don't replace this with anything

File: model/textparse/benchmark_test.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+		parser    string
+
+		compareToExpfmtFormat expfmt.FormatType
+	}{
+		{dataFile: "promtestdata.txt", parser: "promtext", compareToExpfmtFormat: expfmt.TypeTextPlain},
+		{dataFile: "promtestdata.nometa.txt", parser: "promtext", compareToExpfmtFormat: expfmt.TypeTextPlain},
+
+		//  We don't pass compareToExpfmtFormat: expfmt.TypeProtoDelim as expfmt does not support GAUGE_HISTOGRAM:
+		// "expfmt.extractSamples: unknown metric family type GAUGE_HISTOGRAM"
+		// TODO(bwplotka): Fix expfmt?
```
Comment (krajorama) [inline model/textparse/benchmark_test.go:77]: I'd remove the TODO as it's not a to do for this repository and if we ever make statistics of TODO comments , it's misleading.

```suggestion
```

File: model/textparse/benchmark_test.go [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+		{dataProto: createTestProtoBuf(b).Bytes(), parser: "promproto"},
+
+		// We don't pass compareToExpfmtFormat: expfmt.TypeOpenMetrics as expfmt does not support OM exemplars:
+		// text format parsing error in line 19: expected integer as timestamp, got "#"
+		// TODO(bwplotka): Fix expfmt?
```
Comment (krajorama) [inline model/textparse/benchmark_test.go:82]: same as above

File: model/textparse/benchmark_test.go [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+		{dataFile: "promtestdata.txt", parser: "promtext", compareToExpfmtFormat: expfmt.TypeTextPlain},
+		{dataFile: "promtestdata.nometa.txt", parser: "promtext", compareToExpfmtFormat: expfmt.TypeTextPlain},
+
+		//  We don't pass compareToExpfmtFormat: expfmt.TypeProtoDelim as expfmt does not support GAUGE_HISTOGRAM:
+		// "expfmt.extractSamples: unknown metric family type GAUGE_HISTOGRAM"
+		// TODO(bwplotka): Fix expfmt?
+		{dataProto: createTestProtoBuf(b).Bytes(), parser: "promproto"},
+
+		// We don't pass compareToExpfmtFormat: expfmt.TypeOpenMetrics as expfmt does not support OM exemplars:
+		// text format parsing error in line 19: expected integer as timestamp, got "#"
```
Comment (krajorama) [inline model/textparse/benchmark_test.go:81]:
```suggestion
		// text format parsing error in line 19: expected integer as timestamp, got "#".
```

File: model/textparse/benchmark_test.go [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+		dataProto []byte
+		parser    string
+
+		compareToExpfmtFormat expfmt.FormatType
+	}{
+		{dataFile: "promtestdata.txt", parser: "promtext", compareToExpfmtFormat: expfmt.TypeTextPlain},
+		{dataFile: "promtestdata.nometa.txt", parser: "promtext", compareToExpfmtFormat: expfmt.TypeTextPlain},
+
+		//  We don't pass compareToExpfmtFormat: expfmt.TypeProtoDelim as expfmt does not support GAUGE_HISTOGRAM:
+		// "expfmt.extractSamples: unknown metric family type GAUGE_HISTOGRAM"
```
Comment (krajorama) [inline model/textparse/benchmark_test.go:76]:
```suggestion
		// "expfmt.extractSamples: unknown metric family type GAUGE_HISTOGRAM".
```

**Review by krajorama — COMMENTED** [review]
Nice job, I've got some nit comments. This should be easy to adopt to #15115  if it ever gets so far.

File: model/textparse/benchmark_test.go (hunk 2 above)
Comment (bwplotka) [reply model/textparse/benchmark_test.go:151]: I forgot to revert commented skip, but I guess that's fine to do it opt-out, not a biggie. It's easy to decide on what to run with `-bench` regex.

File: model/textparse/promparse_test.go (hunk 3 above)
Comment (bwplotka) [reply model/textparse/promparse_test.go:638]: Nice, added to description. Removed intentionally.

File: model/textparse/benchmark_test.go (hunk 4 above)
Comment (bwplotka) [reply model/textparse/benchmark_test.go:77]: Well, it's our project, so technically it belongs here IF that was a known gap we want to deliver 🤔 

Will replace with concrete issues ID e.g. https://github.com/prometheus/common/issues/430, thanks!

**Review by krajorama — APPROVED** [review]
LGTM

**Pull request merged**